gamemodes/terrortown/entities/entities/ttt_smoke.lua

```diff
@@ -0,0 +1,396 @@
+
+			p:SetCollide(false)
+			p:SetLighting(false)
+			p:SetBounce(0)
+
+			-- bloom the particles out for their last X seconds of life
+			if self:GetFadeTime() > 0 then
+				self.particleData[ #self.particleData + 1 ] = {
+					end_size = math.max(RandomWrap(
+						self:GetEndSizeBase() - self:GetEndSizeVariance(),
+						self:GetEndSizeBase() + self:GetEndSizeVariance()
+					), 0),
+					future_die_time = RandomWrap(
+						0,
+						self:GetFadeTime()
+					)
+				}
+			end
+		end
+	end
+
+	self.Emitter:Finish()
+end
+
+---
+-- Takes a point in space and blows all our particles away from it, disabling the smoke effect.
+-- Moves particles on client, disales on server.
+-- @param Vector pos The point of origin for the effect that will disperse this smoke.
+-- @param number radius The radius of effect for the thing that is dispersing this smoke.
+-- @param number force How strong the effect of the thing that is dispersing this smoke is.
+-- @realm shared
+function ENT:DisperseFromPoint(pos, radius, force)
+	self:Disipate()
+
+	if not CLIENT then return end
+
+	local distSqr = radius * radius
+
+	for i = 1, self:GetDensity() do
+		local particle = self.particles[i]
+		local dist = particle:GetPos():DistToSqr(pos)
+
+		if dist > distSqr then continue end
+
+		local scale = 1 - (dist / distSqr)
+		local dir = (particle:GetPos() - pos):GetNormal()
+		local pushAngle = dir * force * scale
+
+		particle:SetAirResistance(self.airResistance * self.disperseAirResistMultiplier)
+		particle:SetGravity(-physenv.GetGravity() * self.disperseGravityMultiplier)
+		particle:SetVelocity(particle:GetVelocity() + pushAngle)
+	end
+end
+
+---
+-- Allows all smoke particles to die, and disables the active extinguishing effect.
+-- @realm shared
+function ENT:Disipate()
+	self:Deactivate()
+	if not CLIENT or self.debugViz then return end
+
+	for i = 1, self:GetDensity() do
+		local particle = self.particles[i]
+		local particleData = self.particleData[i]
+
+		particle:SetLifeTime(0)
+		particle:SetDieTime(particleData.future_die_time)
+
+		particle:SetAirResistance(self.airResistance)
+		particle:SetBounce(self.bounce)
+		particle:SetCollide(true)
+		particle:SetGravity(-physenv.GetGravity())
+		particle:SetVelocity((self:GetPos() - particle:GetPos()):GetNormal() * self.dispersalVelocity)
+
+		particle:SetEndAlpha(0)
+		particle:SetEndSize(particleData.end_size)
+	end
+end
+
+
+hook.Add("TTT2ConfGrenadeExplode", "TTT2SmokeEntityConfGrenadeExplode", function(pos, radius, push_force)
+	local found = ents.FindInSphere(pos, radius)
+	for _, ent in ipairs(found) do
+		if ent:GetClass() == "ttt_smoke" then
+			ent:DisperseFromPoint(pos, radius, push_force)
+		end
+	end
+end)
+
+---
+-- Begins killing the smoke effect, fades out the looping sound, sets a deactivate time so it can prepare to remove itself.
+-- Only has an effect on the SERVER.
+-- @realm shared
+function ENT:Deactivate()
+	if not SERVER then return end
```

Comment:
```suggestion
	if CLIENT then return end
```
"This does nothing on the CLIENT" is more straightforward than "this does nothing on not the SERVER"